ert/data/record/_record.py

```diff
@@ -0,0 +1,234 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+
+import ert
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+    @staticmethod
+    def create(
+        spec_or_record: Union[
+            Dict[str, bytes], Dict[str, numerical_record_data], "Record"
```

Comment:
Should this not use `record_data` instead of `Dict[str, bytes], Dict[str, numerical_record_data]`? I'm assuming `Union` is associative?

Reply:
This `create` method is essentially a hack the unit tests use. Let's remove it and fix the unit tests?

Reply:
Sounds good 👍🏻